src/Unsafe/Linear.hs

```diff
@@ -20,42 +26,138 @@
+--   :: (a %m-> b -> c %1-> d) %1-> (a %1-> b %1-> c %x-> d)
+-- 'toLinear3' = toLinearN \@3
+-- @
+toLinearN :: forall n f g. ToLinearN n f g => f %1-> g
+toLinearN f = case unsafeLinearityProofN @n @f @g of
+  UnsafeRefl -> f
+
+-- | @ToLinearN n f g@ means that @f@ and @g@ are the same with the
+-- possible exception of the multiplicities of the first @n@ arrows.
+type ToLinearN :: forall {rep :: RuntimeRep}. Nat -> TYPE rep -> TYPE rep -> Constraint
+class ToLinearN n f g where
+  -- | Given that @f@ and @g@ are the same, with the possible exception of the
+  -- multiplicities of the first @n@ arrows, @unsafeLinearityProofN \@n \@f \@g@
+  -- is a fake proof that @f@ and @g@ are identical. This is used primarily in the
+  -- definition of 'toLinearN', but it can also be used, for example, to coerce
+  -- a container of functions:
+  --
+  -- @
+  -- linearMany :: forall a b c. [a -> b -> c] %1-> [a %1-> b %1-> c]
+  -- linearMany = castWithUnsafe (applyUnsafe (UnsafeRefl @[]) $
+  --   unsafeLinearityProofN @2 @(a -> b -> c) @(a %1-> b %1-> c))
+  --
+  -- applyUnsafe :: UnsafeEquality f g -> UnsafeEquality x y -> UnsafeEquality (f x) (g y)
+  -- applyUnsafe UnsafeRefl UnsafeRefl = UnsafeRefl
+  --
+  -- castWithUnsafe :: UnsafeEquality x y -> x %1-> y
+  -- castWithUnsafe UnsafeRefl x = x
+  -- @
+  --
+  -- The rather explicit handling of coercions seems to be necessary,
+  -- unfortunately, presumably due to the way GHC eagerly rejects equality
+  -- constraints it sees as definitely unsatisfiable.
+  unsafeLinearityProofN :: UnsafeEquality f g
+
+instance (ToLinearN' ni f g, ni ~ ToINat n) => ToLinearN n f g where
+  unsafeLinearityProofN = prf @ni
+
+-- | Plain old inductive natural numbers.
+data INat = Z | S INat
+
+-- | Convert a GHC 'Nat' to a real inductive natural number.
+-- We use this because GHC 'Nat' offers a friendly API but
+-- it's a terrible pain for implementation.
+type ToINat :: Nat -> INat
+type family ToINat n where
+  ToINat 0 = 'Z
+  ToINat n = 'S (ToINat (n - 1))
+
+-- | The actual implementation of 'ToLinearN', using the inductive natural
+-- number it's handed.
+type ToLinearN' :: forall {rep :: RuntimeRep}. INat -> TYPE rep -> TYPE rep -> Constraint
+class ToLinearN' arrs f g where
+  prf :: UnsafeEquality f g
+
+instance a ~ b => ToLinearN' 'Z (a :: TYPE rep) (b :: TYPE rep) where
+  prf = UnsafeRefl
+
+-- We use heterogeneous equality here to shift @rep ~ 'LiftedRep@ to the left
+-- side of the fat arrow, so that seeing @'S k@ lets GHC /infer/ that @x@ and
+-- @y@ are lifted types, rather than needing that information to come from
+-- elsewhere.
+instance
+  ( ToLinearN' k fb gb
+  , x ~~ ((a :: TYPE repa) %p-> (fb :: TYPE repb))
+  , y ~~ (a %q-> (gb :: TYPE repb))
+  ) => ToLinearN' ('S k) (x :: TYPE rep) (y :: TYPE rep) where
+  prf = case prf @k @fb @gb of
+    UnsafeRefl -> case unsafeEqualityProof @p @q of
+      UnsafeRefl -> UnsafeRefl
```

Comment:
This is just 
```suggestion
    UnsafeRefl -> unsafeEqualityProof @p @q
```

Isn't it?

Reply:
No, it isn't. We're looking at types

```haskell
a %p-> r1
-- and
a %q-> r2
```

We use a recursive call to get `UnsafeEquality r1 r2`, and we use `unsafeEqualityProof` to get `UnsafeEquality p q`. We need `UnsafeEquality (a %p-> r1) (a %q-> r1)`, and we can't get that by substitution into either of the proofs we have; we need to match on both to build it.

Reply:
Indeed, thanks.